### DISCUSSION TITLE
Question: How to determine declared function names?

### DISCUSSION THREAD

**Discussion by sandreas** [sebastienros/jint#1199] [Q&A] [answered]
Hello,

thank you for jint - great project :-) I would like to determine, which functions are declared in a script. My approach would be to DIFF the existing global function names (like `alert`) with the ones after executing the script.

```c#
var engine = new Engine();
    
engine.Execute(@"
    function hello() { 
        
    };
");
```

Is there any way to get a list that contains `hello` as newly declared function?

Thank you.

**Comment by lahma** [comment, marked as answer]
You need to go through the AST tree produced by Esprima. Get `Script` instance from `JavaScriptParser`'s Parse, then walk it recursive. [Here's an example of that](https://github.com/sebastienros/jint/blob/main/Jint/HoistingScope.cs#L205-L297).

A simpler solution might be just inheriting from `AstVisitor` and override the `VisitFunctionDeclaration`, that will be called for every function.

**Comment by sandreas** [comment]
> You need to go through the AST tree produced by Esprima. Get Script instance from JavaScriptParser's Parse, then walk it recursive. [Here's an example of that](https://github.com/sebastienros/jint/blob/main/Jint/HoistingScope.cs#L205-L297).

Thank you for the quick response. Unfortunately this is way too complex for my use case, but I think I found another solution that might not require this at all.

> A simpler solution might be just inheriting from AstVisitor and override the VisitFunctionDeclaration, that will be called for every function.

That might be a good option, if I cannot work around this problem the other way. Just for the sake of completeness: I thought it would be as easy as this: https://stackoverflow.com/questions/28367663/how-to-get-all-variables-and-functions-names-from-the-global-scope-in-jint

Thank you and keep up the good work.

**Comment by sandreas** [comment]
@lahma I found another possibility. Just for the sake of completeness, I use a combination and function names and global namespace injection for my use case ([`tone`](https://github.com/sandreas/tone) is a C# command line audio tagger):

```js
function musicbrainz(metadata, parameters) {
    // ...
}
tone.RegisterTagger("musicbrainz");
```

- publish `tone` in the global namespace: https://github.com/sandreas/tone/blob/a0a99eb61bf4c44a07d2f8b44513e0721f240ff2/tone/Program.cs#L124
- definition of the published `RegisterTagger` method: https://github.com/sandreas/tone/blob/a0a99eb61bf4c44a07d2f8b44513e0721f240ff2/tone/Services/JavaScriptApi.cs#L43
- `ScriptTagger` class handling the call of the function via `Engine.Invoke`: https://github.com/sandreas/tone/blob/a0a99eb61bf4c44a07d2f8b44513e0721f240ff2/tone/Metadata/Taggers/ScriptTagger.cs#L15


Works pretty good. Now I can create custom audio tagger logic via JavaScript. Thanks for that :-)

For more information see: https://github.com/sandreas/tone#custom-scripted-taggers-experimental
